papers/copper-mountain-2021/scripts/two-grid-hyperelasticity.jl

```diff
@@ -0,0 +1,127 @@
+# dependencies
+using LFAToolkit
+using LinearAlgebra
+using CSV
+using DataFrames
+
+# setup
+dimension = 3
+numbercomponents = 3
+mesh = Mesh3D(1.0, 1.0, 1.0)
+maxeigenvalues = DataFrame()
+
+# constants
+e = 1E6                             # Young's modulus
+ν = 0.3                             # Poisson's ratio
+K = e / (3 * (1 - 2 * ν))           # bulk modulus
+λ = e * ν / ((1 + ν) * (1 - 2 * ν)) # Lamé parameters
+μ = e / (2 * (1 + ν))
+
+# test range
+for fineP = 1:4
+    for coarseP = 0:fineP-1
+        # setup
+        # -- bases
+        coarsep = 2^coarseP
+        finep = 2^fineP
+
+        finebasis = TensorH1LagrangeBasis(finep, finep, numbercomponents, dimension)
+        coarsebasis = TensorH1LagrangeBasis(coarsep, finep, numbercomponents, dimension)
+        ctofbasis = TensorH1LagrangeBasis(
+            coarsep,
+            finep,
+            numbercomponents,
+            dimension,
+            lagrangequadrature = true,
+        )
+
+        # state
+        gradu = [1; 2; 3] * ones(1, 3);
```

Comment:
@jedbrown I seem to remember you mentioning having a suggestion for a particular state vector to use, but I can't seem to find the conversation.

Reply:
Yeah, this is the current gradient of the displacement that we are linearizing about.

I'd like to be able to do the analysis to see if 3rd or 4th order Chebyshev smoothing gives better smoothing and verify with our mini-app, which is why I went with hyperelasticity instead of linear. 

I've been meaning to do a p-convergence study with a twist, so maybe I'll look at that and switch to linear elasticity if becomes difficult to interpret and present the results?

Reply:
The Taylor vortex as a displacement is another (divergence-free) option that can be made periodic. But note that these won't really fit into a single element (except very high order) so you'd need multiple elements. You can do that with macro-elements, but the cost does go up.